// tsyringe/dist/cjs/decorators/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var auto_injectable_1 = require("./auto-injectable");
Object.defineProperty(exports, "autoInjectable", { enumerable: true, get: function () { return auto_injectable_1.default; } });
var inject_1 = require("./inject");
Object.defineProperty(exports, "inject", { enumerable: true, get: function () { return inject_1.default; } });
var injectable_1 = require("./injectable");
Object.defineProperty(exports, "injectable", { enumerable: true, get: function () { return injectable_1.default; } });
var registry_1 = require("./registry");
Object.defineProperty(exports, "registry", { enumerable: true, get: function () { return registry_1.default; } });
var singleton_1 = require("./singleton");
Object.defineProperty(exports, "singleton", { enumerable: true, get: function () { return singleton_1.default; } });
var inject_all_1 = require("./inject-all");
Object.defineProperty(exports, "injectAll", { enumerable: true, get: function () { return inject_all_1.default; } });
var inject_all_with_transform_1 = require("./inject-all-with-transform");
Object.defineProperty(exports, "injectAllWithTransform", { enumerable: true, get: function () { return inject_all_with_transform_1.default; } });
var inject_with_transform_1 = require("./inject-with-transform");
Object.defineProperty(exports, "injectWithTransform", { enumerable: true, get: function () { return inject_with_transform_1.default; } });
var scoped_1 = require("./scoped");
Object.defineProperty(exports, "scoped", { enumerable: true, get: function () { return scoped_1.default; } });
